stl/inc/xutility

```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
-};
+// Matching integer type are eligible for memcopy
+template <class _Source, class _Dest, _Memcopy_cat _Cat,
+    bool = _Memcopy_matching_integral<_Unwrap_enum_t<_Source>, _Unwrap_enum_t<_Dest>>>
+_INLINE_VAR constexpr bool _Can_memcopy_elements = true;
 
-struct _False_copy_cat {
-    static constexpr bool _Really_trivial     = false;
-    static constexpr bool _Trivially_copyable = false;
-};
+// Pointer elements are eligible for memcopy when they point to the same type, ignoring cv-qualification.
+// This handles pointers to object types, pointers to void, and pointers to function types.
+// Performance note: This doesn't attempt to handle `object* == void*`.
+template <class _Source, class _Dest, _Memcopy_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memcopy_elements<_Source*, _Dest*, _Cat, false> =
+    is_same_v<remove_cv_t<_Source>, remove_cv_t<_Dest>>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat : _False_copy_cat {}; // unwrap the pointer-like type and dispatch to _Ptr_cat_helper for copy
+// Types are eligible for memcopy when they are trivially copyable.
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Copy, false> =
+        is_trivially_copyable_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat<_Source*, _Dest*>
-    : conditional_t<is_trivially_assignable_v<_Dest&, _Source&>,
-          _Ptr_cat_helper<remove_cv_t<_Source>, remove_cv_t<_Dest>>, _False_copy_cat> {};
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Move, false> =
+        is_trivially_copyable_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat<move_iterator<_Source*>, _Dest*> : _Ptr_copy_cat<_Source*, _Dest*> {};
+// Types are eligible for memcopy into uninitialized memory when they are trivial.
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Copy_uninitialized, false> =
+        is_trivial_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_move_cat : _False_copy_cat {}; // unwrap the pointer-like type and dispatch to _Ptr_cat_helper for move
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Move_uninitialized, false> =
+        is_trivial_v<_Ty>;
 
+// Arbitrary types are not eligible
+template <class _Source, class _Dest, _Memcopy_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memcopy_elements<_Source, _Dest, _Cat, false> = false;
+
+// Different types are eligible for memcopy when they are trivially copy assignable.
+template <class _Source, class _Dest, _Memcopy_cat>
+_INLINE_VAR constexpr bool _Can_assign_elements = is_trivially_assignable_v<_Dest&, _Source&>;
+
+// Different types are eligible for memmove when they are trivially copy assignable.

```

Comment:
I am confused by this comment - it is referring to `memmove` (which handles overlapping) but the arguments to the type trait indicate that the source is being treated as an rvalue (i.e. C++ moving). I think that the comment saying "trivially copy assignable" should be "trivially move assignable"?

Reply:
"trivially move assignable" was the correct phrase